src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,166 @@
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexForAdd(index);
+        if (index == size) {
+            add(value);  // Додаємо в кінець
+        } else {
+            Node<T> current = findNodeByIndex(index);
+            Node<T> newNode = new Node<>(value, current.prev, current);
+
+            if (current.prev == null) {
+                head = newNode;
+            } else {
+                current.prev.next = newNode;
+            }
+            current.prev = newNode;
+            size++;
+        }
     }
 
     @Override
-    public void addAll(List<T> list) {
+    public boolean addAll(List<T> list) {
+        if (list == null || list.isEmpty()) {
+            return false;
+        }
+        for (T element : list) {
+            add(element);
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return findNodeByIndex(index).data;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> targetNode = findNodeByIndex(index);
+        T oldValue = targetNode.data;
+        targetNode.data = value;
+        return oldValue;
     }
 
+
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> toRemove = findNodeByIndex(index);
+        return unlink(toRemove);
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        while (current != null) {
+            if ((object == null && current.data == null)
+                    || (object != null && object.equals(current.data))) {
+                unlink(current);
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
-    @Override
-    public int size() {
-        return 0;
+    private void checkIndexForAdd(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException(
+                    "Index outside the list: " + index);
+        }
     }
 
-    @Override
-    public boolean isEmpty() {
-        return false;
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException(
+                    "Index outside the list: " + index);
+        }
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        Node<T> current;
+        if (index < size / 2) {
+            current = head;
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+        } else {
+            current = tail;
+            for (int i = size - 1; i > index; i--) {
+                current = current.prev;
+            }
+        }
+        return current;
+    }
+
+    private T unlink(Node<T> node) {
```

Comment:
Create a separate method `unlink(Node node)`. This logic will be used in 2 methods: when we remove from the list by index and by value. So let's create a method that will take a Node that needs to be removed and unlink it. As a result, we will call it from both methods when Node is found.